### PR TITLE
Python Wrapper: Branch and Tag Managers

### LINKED ISSUE #7054: Python Wrapper: Implement Branch and Tag Managers
image

### PR DESCRIPTION
**Pull request by N-o-Z**
Closes #7054

### Background

Implement Branch and Tag Managers

### Testing Details

Added integration tests

### CODE REVIEW

File: clients/python-wrapper/lakefs/reference.py [hunk 1]
```diff
@@ -192,3 +176,28 @@ def __str__(self) -> str:
...
+        while has_more:
+            page = func(*args, **kwargs)
+            has_more = page.pagination.has_more
+            kwargs["after"] = page.pagination.next_offset
+            for res in page.results:
+                yield res
+
+                if max_amount is not None:
+                    max_amount -= 1
+                    if max_amount <= 0:
```
Comment (nopcoder) [inline clients/python-wrapper/lakefs/reference.py:202]: looks like we are still not passing any value to the server to manage the max amount required for listing - can be in a different PR.

File: clients/python-wrapper/lakefs/managers.py
Comment (nopcoder) [inline clients/python-wrapper/lakefs/managers.py:1]: Can we have type for each manager and reuse by composition or calling the generator helper function?
Having a base class and generics and callback for listing will just create complexity.
Check the SDK generated code - it uses the same function to perform the actual call and this call is used by different classes that will verify the function args and prepare the generic call to perform the call.
In this case the two managers tags and branches share the idea of listing - we can use the same function passing callback and get two listing.
At the point where you will have a function just for one manager you start switching to implement methods based on inheritance - revert will go to the lower class, list to the base class. Create to the base and etc...

File: clients/python-wrapper/tests/integration/test_managers.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+from tests.integration.conftest import _setup_repo, get_storage_namespace
+
+
+@pytest.fixture(name="setup_repo_with_branches_and_tags", scope="session")
+def fixture_setup_repo_with_branches_and_tags():
+    clt, repo = _setup_repo(get_storage_namespace("branches-and-tags"),
+                            "branches-and-tags",
+                            "main")
+    for i in range(10):
+        for j in range(100):
```
Comment (nopcoder) [inline clients/python-wrapper/tests/integration/test_managers.py:12]: A lot of branches and tags - we are just checking the API here

File: clients/python-wrapper/tests/integration/test_managers.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+            b = repo.branch(f"branches{i:02d}-{j:03d}").create("main")
+            repo.tag(tag_id=f"tags{i:02d}-{j:03d}").create(b)
+
+    return clt, repo
+
+
+@pytest.mark.parametrize("attr", ("branches", "tags"))
+def test_branch_manager(setup_repo_with_branches_and_tags, attr):
```
Comment (nopcoder) [inline clients/python-wrapper/tests/integration/test_managers.py:20]: Testing X manager?
Still think we can have one test method for each or have two entry points and call the same test method

**Review by nopcoder — CHANGES_REQUESTED** [review]

File: clients/python-wrapper/lakefs/reference.py (hunk 1 above)
Comment (N-o-Z) [reply clients/python-wrapper/lakefs/reference.py:202]: We are passing any additional value via the kwargs - there's no need for any other param

File: clients/python-wrapper/lakefs/managers.py
Comment (N-o-Z) [reply clients/python-wrapper/lakefs/managers.py:1]: Done, though I believe it just creates redundant code taking into account that we are going to have additional managers all with the same logic

File: clients/python-wrapper/tests/integration/test_managers.py (hunk 2 above)
Comment (N-o-Z) [reply clients/python-wrapper/tests/integration/test_managers.py:12]: Not too much :) to check the api we want to check edge cases and pagination so we need more than the default page size

File: clients/python-wrapper/tests/integration/test_managers.py (hunk 3 above)
Comment (N-o-Z) [reply clients/python-wrapper/tests/integration/test_managers.py:20]: Fixed

File: clients/python-wrapper/lakefs/repository.py [hunk 4]
```diff
@@ -118,27 +116,49 @@ def tag(self, tag_id: str) -> Tag:
...
-    def metadata(self) -> dict[str, str]:
+    def branches(self, max_amount: Optional[int] = None,
+                 after: Optional[str] = None, prefix: Optional[str] = None, **kwargs) -> Generator[Branch]:
         """
-        Returns the repository metadata
+        Returns a generator listing for branches on the given repository
+
+        :param max_amount: Stop showing changes after this amount
+        :param after: Return items after this value
+        :param prefix: Return items prefixed with this value
```
Comment (nopcoder) [inline clients/python-wrapper/lakefs/repository.py:126]: doc for kwargs

File: clients/python-wrapper/lakefs/repository.py [hunk 5]
```diff
@@ -118,27 +116,49 @@ def tag(self, tag_id: str) -> Tag:
...
+    def tags(self, max_amount: Optional[int] = None,
+             after: Optional[str] = None, prefix: Optional[str] = None, **kwargs) -> Generator[Tag]:
         """
-        Returns a BranchManager object for this repository
+        Returns a generator listing for tags on the given repository
+
+        :param max_amount: Stop showing changes after this amount
+        :param after: Return items after this value
+        :param prefix: Return items prefixed with this value
+        :raises:
```
Comment (nopcoder) [inline clients/python-wrapper/lakefs/repository.py:145]: kwargs missing

File: clients/python-wrapper/lakefs/reference.py (hunk 1 above)
Comment (nopcoder) [reply clients/python-wrapper/lakefs/reference.py:202]: It is back to the fact that we provide some collection interface and passing max_amount and using it only on client side to limit the iteration, which I think is not required.
We have a different issue for max_amount we can resolve it not in this pr.

File: clients/python-wrapper/tests/integration/test_repository.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    return clt, repo
+
+
+@pytest.mark.parametrize("attr", ("branches", "tags"))
+def test_repository_listings(setup_repo_with_branches_and_tags, attr):
+    _, repo = setup_repo_with_branches_and_tags
+
+    generator = getattr(repo, attr)
+
+    total = 200
```
Comment (nopcoder) [inline clients/python-wrapper/tests/integration/test_repository.py:25]: can we have this one calculated based on the constants used in fixture_setup_repo_with_branches_and_tags?

**Review by nopcoder — APPROVED** [review]
LGTM, the collections implementation using generator works.
There is another way to perform something similar - return a type that implements iterations.

Using this technic the collection can be iterate multiple times without invoking the operation to get the collection.
Both ways works for me - but it is something to consider.

+ kept minor comments.
